rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,298 @@
+1. 'Healthy' (status `True`)
+2. 'Unhealthy' (status `False`)
+3. 'Unknown' (status `Unknown`)
+
+It is up to each template author to determine what "healthy" and "unhealthy" mean for the given resource the template is stamping out. "Unknown" strictly represents that Cartographer has not been able to determine health.
+
+Each template will now have a new field in the spec `healthRule` where authors can specify one of three ways to determine the health of the underlying resource for that template. If no `healthRule` is defined, Cartographer will default to listing the resource as `Healthy` once it has been successfully applied to the cluster and any relevant outputs have been read off the resource.
+
+The `healthRule` requires defining one of three fields:
+1. `alwaysHealthy: {}`
+    * If defined, the resource will always be listed as healthy.
+2. `singleConditionType: <condition type>`
+    * If defined, the health of the resource will be determined by the single condition specified
+      * e.g. if it is set to `singleConditionType: Ready`, the resource will be 'Healthy' if `Ready: True` and 'Unhealthy' if `Ready: False`. Otherwise, it will be 'Unknown'.
+3. `multiMatch`
+    * If defined, matchers for both the 'Healthy' state and the 'Unhealthy' state must be defined.
+    * The matchers in the 'Healthy' state are ANDed together, whereas the matchers in the 'Unhealthy' state are ORed.
+      * That is, a resource is considered to be 'Healthy' only if _all of_ the matchers are fulfilled, but is considered to be 'Unhealthy' if _any one of_ the matchers are fulfilled.
+    * The supported matchers are `matchConditions` and `matchFields`. Each is an array, and an author can define either or both of these matchers for a given state.
+    * For `matchConditions`, authors need only define the `type` and `status` of the condition for Cartographer to read. In the owner status, the full condition will be reflected (with the `reason`, `message`, etc.)
+    * For `matchFields`, the spec is based off of `matchFields` in blueprint resource options. Authors need to define the `key` and `operator`, and potentially the `values` (if the operator is `In` or `NotIn`). Authors can also optionally define a `messagePath`, which is a path in the resource where Cartographer can pull off a message to display in the owner status.
+    
+Each owner's status.resources will now include a `conditions` field (of type []metav1.condition). To make room for future extensibility, this will be an array of conditions, though this RFC only defines one condition to be included here. The condition will be of `type` 'Healthy' and will report a `status` of 'True', 'False', or 'Unknown'. The condition will also have a `reason` which will be based off the `healthRule` as follows:
+    
+* If `alwaysHealthy: {}`, `reason: AlwaysHealthy`.
+* If `singleConditionType: X`, `reason: XCondition`.
+* If `multiMatch: ...`, `reason: (MatchedCondition|MatchedField)`.
```

Comment:
Consider the case of a multimatch with a healthy matchCondition and unhealthy matchField. When status is unknown, what would be the reason? (Reason is a required field and may not be an empty string)

Reply:
unhealthy signals are more prominent than unknown signals, unknown signals are more prominent than healthy signals

Reply:
Interesting question. As described currently in the RFC, it would be the first one Cartographer tries to read:
`In the case of multiMatch, if there is more than one matching condition or field that caused the resource to enter it's current state, the first one read will be propagated by Cartographer.`

But perhaps we could have a special reason for Unknown in multiMatch?

Reply:
> But perhaps we could have a special reason for Unknown in multiMatch? What do you think?

I like that idea. Something along the lines of `NoMatchesFulfilled` strikes me as most informative in this case.

Reply:
Sounds good to me. Updated the RFC with this suggestion.